File: union-find-forest/union_find_forest.py
class DSU(object):
    def __init__(self, n):
        self.array = [i for i in range(n)]

    
    def find(self, x):
        while (self.array[x] != x):
            x = self.array[x]
        return x


    def union(self, first, second):
        x = self.find(first)
        y = self.find(second) 
        if(x == y):
            return "YES"
        else:
            self.array[x] = y
            return "NO"


n, k = [int(x) for x in input().split()]
dsu = DSU(n)
for _ in range(n):
        first, second = [int(x) for x in input().split()]
        res = dsu.union(first, second)
        print(res)

File: union-find-forest/src/UnionFindForest.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class UnionFindForest {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));

        String[] input = (reader.readLine().split(" "));
        Forest forest = new Forest(Integer.parseInt(input[0]));

        for (int i = 0; i < Integer.parseInt(input[1]); i++) {
            String[] nums = (reader.readLine().split(" "));
            if (forest.union(Integer.parseInt(nums[0]), Integer.parseInt(nums[1]))) {
                System.out.println("YES");
            } else {
                System.out.println("NO");
            }
        }
    }

    public static class Forest {

        private int[] forest;

        public Forest(int size) {
            forest = new int[size];
            for (int i = 0; i < size; i++) {
                forest[i] = i;
            }
        }

        public int get(int x) {
            if (forest[x] != x) {
                forest[x] = get(forest[x]);
            }
            return forest[x];
        }

        public boolean union(int x, int y) {
            x = get(x);
            y = get(y);

            if (x != y) {
                forest[x] = y;
                return false;
            }

            return true;
        }

    }
}
